man/man8/gethostlatency.8
.TH gethostlatency 8  "2016-01-28" "USER COMMANDS"
.SH NAME
gethostlatency \- Show latency for getaddrinfo/gethostbyname[2] calls. Uses Linux eBPF/bcc.
.SH SYNOPSIS
.B gethostlatency
.SH DESCRIPTION
This traces and prints when getaddrinfo(), gethostbyname(), and gethostbyname2()
are called, system wide, and shows the responsible PID and command name,
latency of the call (duration) in milliseconds, and the host string.

This tool can be useful for identifying DNS latency, by identifying which
remote host name lookups were slow, and by how much.

This makes use of a Linux 4.4 feature (bpf_perf_event_output());
for kernels older than 4.4, see the version under tools/old,
which uses an older mechanism

This tool currently uses dynamic tracing of user-level functions and registers,
and may need modifications to match your software and processor architecture.

Since this uses BPF, only the root user can use this tool.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\-p PID
Trace this process ID only.
.SH EXAMPLES
.TP
Trace host lookups (getaddrinfo/gethostbyname[2]) system wide:
#
.B gethostlatency
.SH FIELDS
.TP
TIME
Time of the command (HH:MM:SS).
.TP
PID
Process ID of the client performing the call.
.TP
COMM
Process (command) name of the client performing the call.
.TP
HOST
Host name string: the target of the lookup.
.SH OVERHEAD
The rate of lookups should be relatively low, so the overhead is not expected
to be a problem.
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Brendan Gregg
.SH SEE ALSO
tcpdump(8)


debian/changelog
bcc (0.12.0-1) unstable; urgency=low

  * Support for kernel up to 5.4
  * klockstat tool to track kernel mutex lock statistics
  * cmake option CMAKE_USE_LIBBPF_PACKAGE to build a bcc shared library
    linking with distro libbpf_static.a
  * new map.lookup_or_try_init() API to remove hidden return in
    map.lookup_or_init()
  * BPF_ARRAY_OF_MAPS and BPF_HASH_OF_MAPS support
  * support symbol offset for uprobe in both C++ and python API,
    kprobe already has the support
  * bug fixes for trace.py, tcpretrans.py, runqslower.py, etc.

 -- Yonghong Song <ys114321@gmail.com>  Tue, 10 Dec 2019 17:00:00 +0000

bcc (0.11.0-1) unstable; urgency=low

  * Support for kernel up to 5.3
  * Corresponding libbpf submodule release is v0.0.5
  * Fix USDT issue with multi-threaded applications
  * Fixed the early return behavior of lookup_or_init
  * Support for nic hardware offload
  * Fixed and Enabled Travis CI
  * A lot of tools change with added new options, etc.

 -- Yonghong Song <ys114321@gmail.com>  Tue, 03 Oct 2019 17:00:00 +0000

bcc (0.10.0-1) unstable; urgency=low

  * Support for kernel up to 5.1
  * corresponding libbpf submodule release is v0.0.3
  * support for reading kernel headers from /proc
  * libbpf.{a,so} renamed to libcc_bpf.{a,so}
  * new common options for some tools
  * new tool: drsnoop
  * s390 USDT support

 -- Brenden Blanco <bblanco@gmail.com>  Tue, 28 May 2019 17:00:00 +0000

bcc (0.9.0-1) unstable; urgency=low

  * Adds support for BTF
  * Uses libbpf common library to wrap syscall API
  * Many bugfixes and new tools

 -- Brenden Blanco <bblanco@gmail.com>  Thu, 07 Mar 2019 17:00:00 +0000

bcc (0.8.0-1) unstable; urgency=low

  * Support for kernel up to 5.0

 -- Brenden Blanco <bblanco@gmail.com>  Fri, 11 Jan 2019 17:00:00 +0000

bcc (0.7.0-1) unstable; urgency=low

  * Support for kernel up to 4.18

 -- Brenden Blanco <bblanco@gmail.com>  Tue, 04 Sep 2018 17:00:00 +0000

bcc (0.6.1-1) unstable; urgency=low

  * Build support for Fedora 28 and Ubuntu 18.04
  * Add option to change license
  * Optimizations for some uses of bpf_probe_reads

 -- Brenden Blanco <bblanco@gmail.com>  Mon, 23 Jul 2018 17:00:00 +0000

bcc (0.6.0-1) unstable; urgency=low

  * Support for kernel up to 4.17
  * Many bugfixes
  * Many new tools
  * Improved python3 support

 -- Brenden Blanco <bblanco@gmail.com>  Wed, 13 Jun 2018 17:00:00 +0000

bcc (0.5.0-1) unstable; urgency=low

  * Support for USDT in ARM64
  * Bugfixes for 4.14 in some tools
  * Fixes for smoke test failures
  * Runtime memory usage reductions

 -- Brenden Blanco <bblanco@gmail.com>  Wed, 29 Nov 2017 17:00:00 +0000

bcc (0.4.0-1) unstable; urgency=low

  * Bugfixes
  * Support for kernel up to 4.14

 -- Brenden Blanco <bblanco@gmail.com>  Fri, 20 Oct 2017 17:00:00 +0000

bcc (0.3.0-1) unstable; urgency=low

  * Many bugfixes
  * Many tools converted to perf ring buffer
  * New utilities in tools/
   * capable, cpuunclaimed, dbslower, dbstat, deadlock_detector, llcstat,
     mountsnoop, runqlen, slabratetop, syscount, tcplife, tcptop, ttysnoop,
     ucalls, uflow, ugc, uobjnew, ustat, uthreads
  * New C++ API
  * Support for kernel up to 4.10

 -- Brenden Blanco <bblanco@gmail.com>  Thu, 09 Mar 2017 19:08:08 +0000

bcc (0.2.0-1) unstable; urgency=low

  * Add many new utilities in tools/
  * Support for USDT
  * Support for lua
  * Many utilities converted to perf ring buffer
  * Support for tracepoints

 -- Brenden Blanco <bblanco@plumgrid.com>  Thu, 08 Sep 2016 17:05:28 -0700

bcc (0.1.8-1) unstable; urgency=low

  * Add many new utilities in tools/
   * wakeuptime, offwaketime, argdist, {xfs,zfs,ext4}{slower,dist}, others
  * Support for bpf_perf_event()
  * Support for public tables shared between programs
  * Support for up to 4.4 features
  * Remove external file dependencies from clang lib

 -- Brenden Blanco <bblanco@plumgrid.com>  Mon, 23 Feb 2016 00:41:00 +0000

bcc (0.1.7-1) unstable; urgency=low

  * Tracing features and bugfixes
  * Built against LLVM 3.8 HEAD

 -- Brenden Blanco <bblanco@plumgrid.com>  Mon, 12 Oct 2015 16:47:09 +0000

bcc (0.1.6-1) unstable; urgency=low

  * Stability fixes
  * Improvements to python API
  * Tracing features
  * Support for kernel 4.2 features

 -- Brenden Blanco <bblanco@plumgrid.com>  Wed, 02 Sep 2015 16:23:19 +0000

bcc (0.1.5-1) unstable; urgency=low

  * Initial release

 -- Brenden Blanco <bblanco@plumgrid.com>  Mon, 06 Jul 2015 18:04:28 +0000


tools/gethostlatency.py
#!/usr/bin/python
#
# gethostlatency  Show latency for getaddrinfo/gethostbyname[2] calls.
#                 For Linux, uses BCC, eBPF. Embedded C.
#
# This can be useful for identifying DNS latency, by identifying which
# remote host name lookups were slow, and by how much.
#
# This uses dynamic tracing of user-level functions and registers, and may
# need modifications to match your software and processor architecture.
#
# Copyright 2016 Netflix, Inc.
# Licensed under the Apache License, Version 2.0 (the "License")
#
# 28-Jan-2016    Brendan Gregg   Created this.
# 30-Mar-2016   Allan McAleavy updated for BPF_PERF_OUTPUT

from __future__ import print_function
from bcc import BPF
from time import strftime
import argparse

examples = """examples:
    ./gethostlatency           # time getaddrinfo/gethostbyname[2] calls
    ./gethostlatency -p 181    # only trace PID 181
"""
parser = argparse.ArgumentParser(
    description="Show latency for getaddrinfo/gethostbyname[2] calls",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-p", "--pid", help="trace this PID only", type=int,
    default=-1)
parser.add_argument("--ebpf", action="store_true",
    help=argparse.SUPPRESS)
args = parser.parse_args()

# load BPF program
bpf_text = """
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

struct val_t {
    u32 pid;
    char comm[TASK_COMM_LEN];
    char host[80];
    u64 ts;
};

struct data_t {
    u32 pid;
    u64 delta;
    char comm[TASK_COMM_LEN];
    char host[80];
};

BPF_HASH(start, u32, struct val_t);
BPF_PERF_OUTPUT(events);

int do_entry(struct pt_regs *ctx) {
    if (!PT_REGS_PARM1(ctx))
        return 0;

    struct val_t val = {};
    u32 pid = bpf_get_current_pid_tgid();

    if (bpf_get_current_comm(&val.comm, sizeof(val.comm)) == 0) {
        bpf_probe_read(&val.host, sizeof(val.host),
                       (void *)PT_REGS_PARM1(ctx));
        val.pid = bpf_get_current_pid_tgid();
        val.ts = bpf_ktime_get_ns();
        start.update(&pid, &val);
    }

    return 0;
}

int do_return(struct pt_regs *ctx) {
    struct val_t *valp;
    struct data_t data = {};
    u64 delta;
    u32 pid = bpf_get_current_pid_tgid();

    u64 tsp = bpf_ktime_get_ns();

    valp = start.lookup(&pid);
    if (valp == 0)
        return 0;       // missed start

    bpf_probe_read(&data.comm, sizeof(data.comm), valp->comm);
    bpf_probe_read(&data.host, sizeof(data.host), (void *)valp->host);
    data.pid = valp->pid;
    data.delta = tsp - valp->ts;
    events.perf_submit(ctx, &data, sizeof(data));
    start.delete(&pid);
    return 0;
}
"""
if args.ebpf:
    print(bpf_text)
    exit()

b = BPF(text=bpf_text)
b.attach_uprobe(name="c", sym="getaddrinfo", fn_name="do_entry", pid=args.pid)
b.attach_uprobe(name="c", sym="gethostbyname", fn_name="do_entry",
                pid=args.pid)
b.attach_uprobe(name="c", sym="gethostbyname2", fn_name="do_entry",
                pid=args.pid)
b.attach_uretprobe(name="c", sym="getaddrinfo", fn_name="do_return",
                   pid=args.pid)
b.attach_uretprobe(name="c", sym="gethostbyname", fn_name="do_return",
                   pid=args.pid)
b.attach_uretprobe(name="c", sym="gethostbyname2", fn_name="do_return",
                   pid=args.pid)

# header
print("%-9s %-6s %-16s %10s %s" % ("TIME", "PID", "COMM", "LATms", "HOST"))

def print_event(cpu, data, size):
    event = b["events"].event(data)
    print("%-9s %-6d %-16s %10.2f %s" % (strftime("%H:%M:%S"), event.pid,
        event.comm.decode('utf-8', 'replace'), (float(event.delta) / 1000000),
        event.host.decode('utf-8', 'replace')))

# loop with callback to print_event
b["events"].open_perf_buffer(print_event)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
